### PR TITLE
[fix] [test] Fix flaky test `MetadataStoreStatsTest.testBatchMetadataStoreMetrics`

### LINKED ISSUE #19792: Flaky-test: org.apache.pulsar.broker.stats.MetadataStoreStatsTest.testBatchMetadataStoreMetrics
### Search before asking

- [X] I searched in the [issues](https://github.com/apache/pulsar/issues) and found nothing similar.

### Example failure

https://github.com/apache/pulsar/actions/runs/4383540966/jobs/7674179427

### Exception stacktrace

i don't known where to find unit test fail stacktrace. please let me know.

### Are you willing to submit a PR?

- [X] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by lifepuzzlefun**
Fixes #19792

### Motivation
Fix flaky test `MetadataStoreStatsTest.testBatchMetadataStoreMetrics`

some unit test create MetadataStore without setting name and not close will cause prometheus metric not unregisterd. 
the unittest above will check promethues metric tag name not null.

### Modifications

1. set metadata store name to avoid generate metric with name cause relate unit test fail.

2. auto generate metadata store name if caller not specify one for avoid this problem occur again.

### Verifying this change

This change is already covered by existing tests, such as *(please describe tests)*.

### Does this pull request potentially affect one of the following parts:

*If the box was checked, please highlight the changes*

- [ ] Dependencies (add or upgrade a dependency)
- [ ] The public API
- [ ] The schema
- [ ] The default values of configurations
- [ ] The threading model
- [ ] The binary protocol
- [ ] The REST endpoints
- [ ] The admin CLI options
- [ ] The metrics
- [ ] Anything that affects deployment

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository:

### CODE REVIEW

**Comment by lifepuzzlefun** [comment]
the auto generate metadata store name will be like `org.apache.pulsar.metadata.MetadataStoreConfigTest_testGeneCallerStackNameWhenMetadataStoreNameNotSet_11`

**Comment by lifepuzzlefun** [comment]
@lhotari please take a look.

**Comment by nodece** [comment]
Is the root cause that the metadatastoreName is empty?

**Comment by lifepuzzlefun** [comment]
> Is the root cause that the metadatastoreName is empty?

@nodece The root cause is create one MetadataStore with no name set and not close it.

File: pulsar-metadata/src/main/java/org/apache/pulsar/metadata/api/MetadataStoreConfig.java [hunk 1]
```diff
@@ -81,6 +84,39 @@ public class MetadataStoreConfig {
...
+        @Override
+        public MetadataStoreConfigBuilder metadataStoreName(String metadataStoreName) {
+            this.callerSetMetadataStoreName = metadataStoreName;
+            return super.metadataStoreName(metadataStoreName);
+        }
+
+        @Override
+        public MetadataStoreConfig build() {
+            if (Strings.isEmpty(this.callerSetMetadataStoreName)) {
+                Exception e = new Exception("stack trace");
```
Comment (nodece) [inline pulsar-metadata/src/main/java/org/apache/pulsar/metadata/api/MetadataStoreConfig.java:104]: I think you can throw an exception, so like:
```
throw new IllegalArgumentException("metadataStoreName is required");
```

Comment (lifepuzzlefun) [reply pulsar-metadata/src/main/java/org/apache/pulsar/metadata/api/MetadataStoreConfig.java:104]: I agree with you. the problem is there is a lot of ut didn't set name. If they don't set name but they close the object the flaky test won't occur.

Comment (nodece) [reply pulsar-metadata/src/main/java/org/apache/pulsar/metadata/api/MetadataStoreConfig.java:104]: So can we just need to close the metadatastore to fix the flaky test?

Comment (lifepuzzlefun) [reply pulsar-metadata/src/main/java/org/apache/pulsar/metadata/api/MetadataStoreConfig.java:104]: yes, other modification is for this case not happen again and supply some method to find the wrong usage.

the modification won't hurt production.

Comment (nodece) [reply pulsar-metadata/src/main/java/org/apache/pulsar/metadata/api/MetadataStoreConfig.java:104]: > yes, other modification is for this case not happen again and supply some method to find the wrong usage.

Makes sense! But this is rare.

Comment (lifepuzzlefun) [reply pulsar-metadata/src/main/java/org/apache/pulsar/metadata/api/MetadataStoreConfig.java:104]: yes i agree with you. maybe we can add one comment in this unit test point to the current modification to catch up new wrong usage. and the branch will be retain for those who is block by this problem.

I'll remove the current metadataStore name generate code for this patch to merge.

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/stats/MetadataStoreStatsTest.java [hunk 2]
```diff
@@ -69,6 +69,9 @@ protected void cleanup() throws Exception {
 
     @Test
     public void testMetadataStoreStats() throws Exception {
+        // if this test fail see link below to find something.
```
Comment (nodece) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/stats/MetadataStoreStatsTest.java:72]: It looks unnecessary.

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/zookeeper/MultiBrokerMetadataConsistencyTest.java [hunk 3]
```diff
@@ -43,6 +44,9 @@ protected int numberOfAdditionalBrokers() {
 
     TestZKServer testZKServer;
 
+    private final ConcurrentHashMap<MetadataStoreExtended, Object> needCloseStore =
```
Comment (nodece) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/zookeeper/MultiBrokerMetadataConsistencyTest.java:47]: Could you use the List?

Comment (lifepuzzlefun) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/zookeeper/MultiBrokerMetadataConsistencyTest.java:47]: will the test run concurrently？

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/stats/MetadataStoreStatsTest.java (hunk 2 above)
Comment (lifepuzzlefun) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/stats/MetadataStoreStatsTest.java]: fixed

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/zookeeper/MultiBrokerMetadataConsistencyTest.java (hunk 3 above)
Comment (lifepuzzlefun) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/zookeeper/MultiBrokerMetadataConsistencyTest.java]: fixed

**Review by nodece — APPROVED** [review]

**Review by coderzc — APPROVED** [review]

**Comment by nodece** [comment]
This fix is invalid, see https://github.com/apache/pulsar/actions/runs/4424020709/jobs/7763604452#step:11:1277

**Pull request merged**